MoreLinq/CountBy.cs

```diff
@@ -84,23 +84,31 @@ void Loop(IEqualityComparer<TKey> cmp)
                     var havePrevKey = false;
                     var prevKey = default(TKey);
                     var index = 0;
+                    var nullIndex = (int?) null;
+
+                    int? IndexOf(TKey key) =>
+                        key == null ? nullIndex is int ni ? ni : (int?) null
+                                    : dic.TryGetValue(key, out var i) ? i : (int?) null;
 
                     foreach (var item in source)
                     {
                         var key = keySelector(item);
 
                         if (// key same as the previous? then re-use the index
-                            (havePrevKey && cmp.GetHashCode(prevKey) == cmp.GetHashCode(key)
-                                          && cmp.Equals(prevKey, key))
+                            havePrevKey && cmp.GetHashCode(prevKey) == cmp.GetHashCode(key)
+                                         && cmp.Equals(prevKey, key)
                             // otherwise try & find index of the key
-                            || dic.TryGetValue(key, out index))
+                            || IndexOf(key) is int i && (index = i) >= 0 /* always true */)
```

Comment:
I think I prefer this to be folded into `IndexOf`: 

```csharp
bool TryGetIndex(TKey key, out int index) {
  if (key == null) {
    index = nullIndex.GetValueOrDefault();
    return nullIndex.HasValue;
  }
  else {
    return dic.TryGetValue(key, out index);
  }
}
```

Assignments within `if` statements always feel icky to me (perhaps this is C-caused "trauma"?)